Authorization cloud mail

kalash9747 opened the pull request:
Добавил модули util и cloudmail.
Хранение пользовательских данных в зашифрованном виде в отдельном файле.
Добавил класс для упрощения отправки Api запросов клиентом java 11 .
Добавил HttpResponseDecorator  для упрощения работы с ответом
Тесты:
      -api авторизация в облаке Mail.ru.
      -проверка присутствия имен файлов корневого каталога в теле ответа .








![image](https://user-images.githubusercontent.com/82981820/168490523-c0177b6c-f34d-4fd9-a190-c8f8d80ebd7b.png)

kalash9747 committed:
add Crypthographer
add UserCryptographer
add MailUserRole
add Column
add Cars

kalash9747 committed:
changed Crypthographer

kalash9747 committed:
changed Crypthographer

kalash9747 committed:
add comments

kalash9747 committed:
add comments

kalash9747 committed:
add comments

kalash9747 committed:
add comments

kalash9747 committed:
Merge remote-tracking branch 'origin/authorizationCloudMail' into authorizationCloudMail

# Conflicts:
#	.gitignore

kalash9747 committed:
add README and TESTRUN

kalash9747 committed:
Merge branch 'master' into authorizationCloudMail

kalash9747 committed:
add models response api/v2/feed
add request api/v2/feed
add HttpResponseDecorator

kalash9747 committed:
delete unused files

kalash9747 committed:
Rename HttpResponseDecorator -> Facade

dKusakin commented on build.gradle:
```diff
@@ -1,30 +1,25 @@
-test {
-    useJUnitPlatform()
+    tasks.withType(JavaCompile) {
+        sourceCompatibility = JavaVersion.VERSION_11
```
Не очень корректно внутри `tasks` указывать версию. Эти параметры чуть глобальнее, указывай хотя бы просто внутри `subprojects`

kalash9747 replied:
сделал

dKusakin commented on cloudmail/build.gradle:
```diff
@@ -0,0 +1,27 @@
+
+allure {
+    autoconfigure = true
+    version = '2.15.0'
```
В куче мест одни те же версии одних и тех же, по сути, ядер артефактов.
А когда обновлять решишься про все вспомнишь?
Хотя бы в корневом `build.gradle` в секции `ext` заведи под них переменные

kalash9747 replied:
сделал

dKusakin commented on cloudmail/src/main/java/api/CloudMailApi.java:
```diff
@@ -0,0 +1,66 @@
+     * Авторизоваться в cloud.mail.ru
+     */
+    public HttpResponseFacade login(User user) {
+        return step("Авторизоваться под пользователем " + user.getLogin(), () -> {
```
А почему степ не аннотацией, а внутри метода?

kalash9747 replied:
сделал

dKusakin commented on cloudmail/src/main/java/api/CloudMailApi.java:
```diff
@@ -0,0 +1,66 @@
+        return step("Авторизоваться под пользователем " + user.getLogin(), () -> {
+            baseUri = user.getUrl();
+            URI authUri = URI.create("https://auth.mail.ru/cgi-bin/auth");
+            String authBody = String.format("username=%s&Login=%s&password=%s&Password=%s&saveauth=1&act_token=%s&page=%s",
```
Хм, как-то не очень выглядит. Мб замутить свой `BodyPublisher` для `form-param`?
Это, конечно, скорее будет а-ля адаптер к `StringPublisher`, но всё равно.
Мне просто не очень нравится, что ты в тело просто `String` пихаешь. Логичнее туда как раз объект поинтереснее класть. Но это больше на подумать, по факту для начала и так хорошо)

kalash9747 replied:
сделал

dKusakin commented on cloudmail/src/main/java/api/CloudMailApi.java:
```diff
@@ -0,0 +1,66 @@
+                .stream()
+                .filter(x -> x.getName().startsWith("act"))
+                .findFirst();
+        return act.isPresent() ? act.get().getValue() : "";
```
После `findFirst()`:
```
    .map(HttpCookie::getValue)
    .orElse("");
```

kalash9747 replied:
крутяк) сделал

dKusakin commented on cloudmail/src/main/java/api/CloudMailApi.java:
```diff
@@ -0,0 +1,66 @@
+
+    @Step("Отправить запрос api/v2/feed")
+    public HttpResponseFacade feed() {
+        return base.get(URI.create(baseUri + "/api/v2/feed"),
```
Кстати, `feed` не совсем по заданию. Он получает то ли вообще все, то ли недавние, но суть в том, что из подпапок тоже достает

kalash9747 replied:
сделал

dKusakin commented on cloudmail/src/test/java/api/HomeDirectoryTest.java:
```diff
@@ -0,0 +1,51 @@
+            add("спиннер.png");
+        }};
+
+        var actualFileNames = cloudMailApi.feed()
```
`var`  - тоже холиварная тема)
Главное - не злоупотреблять

kalash9747 replied:
а он вообще затесался случайно, сделал

dKusakin commented on cloudmail/src/test/java/api/HomeDirectoryTest.java:
```diff
@@ -0,0 +1,51 @@
+import static java.util.stream.Collectors.toSet;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class HomeDirectoryTest {
```
Хоть я этого и не указал, но, если не сложно, мб будем разлогиниваться?

kalash9747 replied:
сделал, проверил - работает , на попытку снова кинуть запрос прилетает ответ неавторизован

dKusakin commented on util/build.gradle:
```diff
@@ -0,0 +1,8 @@
+dependencies {
+    implementation 'com.fasterxml.jackson.core:jackson-databind:2.13.2'
```
Оставлю про версии тут тоже, чтобы не забыть посмотреть

kalash9747 replied:
сделал

dKusakin commented on util/src/main/java/encryption/UserRole.java:
```diff
@@ -0,0 +1,7 @@
+package encryption;
+
+/**
+ * Интерфейс рпользовательских ролей
```
**р**пользовательских

kalash9747 replied:
сделал

kalash9747 committed:
add buildSrc/../Deps
add StringPublisherAdapter
add QueryParam
add logout()
change query feed()->privateList()
add models for api/v4/private/list?path=...

dKusakin commented on cloudmail/src/main/java/api/CloudMailApi.java:
```diff
@@ -0,0 +1,78 @@
+        baseUri = user.getUrl();
+        URI authUri = URI.create(baseAuthUri + "/auth");
+        BodyPublisher authBody = ofQueryParams(
+                new QueryParam("username", user.getLogin()),
```
А я думал мапой замутить. Тут бы через `Map.of()` (11-я, же джава, можем себе позволить) попроще смотрелось бы. Просто вариант на перегрузку

dKusakin commented on util/src/main/java/api/QueryParam.java:
```diff
@@ -0,0 +1,20 @@
+/**
+ * Параметр для Api запросов
+ */
+public class QueryParam {
```
В принципе, в качестве этой штуки мог бы выступать и `AbstractMap.SimpleEntry`

kalash9747 committed:
ObjectMapper FAIL_ON_UNKNOWN_PROPERTIES = false
add gradle.properties

kalash9747 committed:
add ofMapEntries
